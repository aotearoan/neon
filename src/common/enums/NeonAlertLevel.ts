/**
 * Alert levels are the severity levels to be used with <a href="/feedback/alert#examples">NeonAlert & NeonToast</a>
 * components to inform the user of the severity of a message.
 */
export enum NeonAlertLevel {
  /** Information level, use for standard informational messages. */
  Info = 'info',
  /** Success level, use for indicating successful actions. */
  Success = 'success',
  /** Warning level, use for indicating a problem that the user <strong>can</strong> recover from. */
  Warn = 'warn',
  /** Error level, use for indicating a problem that the user <strong>cannot</strong> recover from. */
  Error = 'error',
}
