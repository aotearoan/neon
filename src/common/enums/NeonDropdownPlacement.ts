/**
 * Describes placement of dropdown related components like:
 * <ul class="no-style">
 *  <li><a href="/user-input/select">NeonSelect</a></li>
 *  <li><a href="/navigation/dropdown-menu">NeonDropdownMenu</a></li>
 *  <li><a href="/presentation/dropdown">NeonDropdown</a></li>
 * </ul>
 *
 * <p>The placement has a <strong>primary placement</strong> (the first direction) & a <strong>secondary placement</strong>
 * (the second direction). The primary placement is the location of the popup relative to the triggering element (e.g.
 * the select button). With the secondary placement defining the alignment relative to the trigger. E.g. BottomLeft will
 * open the dropdown below the trigger element with the left-hand edges aligned.</p>
 *
 * <p>The placement utilities are smart & dynamic. They listen to resize events & aim for best effort based on where the
 * dropdown is opened. E.g. in the case above with BottomLeft if there is not enough visible room to align the component
 * to the left then it will try to align to the right & likewise if there is not enough space below the trigger it will
 * align to the top (& reverse the option order in the case of a select).</p>
 */
export enum NeonDropdownPlacement {
  /** Placement above trigger element, aligned to the left */
  TopLeft = 'top-left',
  /** Placement above trigger element, aligned to the right */
  TopRight = 'top-right',
  /** Placement below trigger element, aligned to the left */
  BottomLeft = 'bottom-left',
  /** Placement below trigger element, aligned to the right */
  BottomRight = 'bottom-right',
  /** Placement to the left of the trigger element, aligned to the top */
  LeftTop = 'left-top',
  /** Placement to the left of the trigger element, aligned to the bottom */
  LeftBottom = 'left-bottom',
  /** Placement to the right of the trigger element, aligned to the top */
  RightTop = 'right-top',
  /** Placement to the right of the trigger element, aligned to the bottom */
  RightBottom = 'right-bottom',
}
