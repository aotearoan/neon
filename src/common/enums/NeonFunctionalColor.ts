/**
 * The core color palettes available for styling components in Neon. See the design section
 * <a href="/design/color">Color</a> page for more details on color palettes.
 */
export enum NeonFunctionalColor {
  /** Low contrast is a more subtle neutral palette which can be used for secondary actions & supporting text. */
  LowContrast = 'low-contrast',
  /** Neutral is the base greyscale palette. */
  Neutral = 'neutral',
  /**
   * High contrast is a higher contrast palette than can be used for increased accessibility or for a stronger black &
   * white application.
   * */
  HighContrast = 'high-contrast',
  /** Brand is the main brand color palette, this is useful for branded CTAs, text & controls. */
  Brand = 'brand',
  /** Primary is the main application color palette. For most applications this will be used by default. */
  Primary = 'primary',
  /** Info can be used to convey informational context, e.g. notes & alerts */
  Info = 'info',
  /** Success can be used to convey successful actions. */
  Success = 'success',
  /** Warn can be used to convey issues/actions that are less severe but may still impact the user. */
  Warn = 'warn',
  /** Error can be used to convey issues/actions that are severe & that the user may not be able to recover from. */
  Error = 'error',
}
