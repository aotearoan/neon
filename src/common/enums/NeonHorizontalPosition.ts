/**
 * Describes the horizontal positioning of certain elements in the context of a component. An example would be an icon
 * in a button or label as well as the label for a checkbox or switch.
 */
export enum NeonHorizontalPosition {
  /** Left horizontal positioning. */
  Left = 'left',
  /** Right horizontal positioning. */
  Right = 'right',
}
