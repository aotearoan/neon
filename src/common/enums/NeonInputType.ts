/**
 * Corresponds to the <a href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#input_types">HTML input
 * types</a> for use with <a href="/user-input/input">NeonInput</a>.
 *
 * NOTE: prefer the specific components for Number, Password, Date, File, Search fields.
 */
export enum NeonInputType {
  /** Date input for selecting dates from a calendar. Use <a href="/user-input/date-picker">NeonDatePicker</a>. */
  Date = 'date',
  /** Date time input for selecting dates including a specific time. */
  DateTime = 'datetime-local',
  /** Email input for entering email addresses. */
  Email = 'email',
  /**
   * File input for opening a file dialog. Use <a href="/user-input/file">NeonDatePicker</a> or
   * <a href="/user-input/drop-zone">NeonDropZone</a>.
   */
  File = 'file',
  /** Number input for entering numeric values. Use <a href="/user-input/number">NeonNumber</a> */
  Number = 'number',
  /** Password input for entering passwords. Use <a href="/user-input/password">NeonPassword</a> */
  Password = 'password',
  /**
   * Range input for a slider control. Use <a href="/user-input/slider">NeonSlider</a> or
   * <a href="/user-input/range-slider">NeonRangeSlider</a>
   */
  Range = 'range',
  /** Search input for searching. Use <a href="/user-input/search">NeonSearch</a> */
  Search = 'search',
  /** Telephone number input. */
  Tel = 'tel',
  /** Text input for single or multiline text input. */
  Text = 'text',
  /** Time input. */
  Time = 'time',
  /** URL input for entering web addresses. */
  Url = 'url',
}
