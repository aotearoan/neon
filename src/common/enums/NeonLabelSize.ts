/**
 * The size options for the <a href="/presentation/label">NeonLabel</a> component.
 */
export enum NeonLabelSize {
  /** Extra extra small, default is 16 rem. */
  ExtraExtraSmall = 'xxs',
  /** Extra small, default is 20 rem. */
  ExtraSmall = 'xs',
  /** Small, default is 28 rem. */
  Small = 's',
  /** Medium, default is 36 rem. */
  Medium = 'm',
  /** Large, default is 44 rem. */
  Large = 'l',
}
