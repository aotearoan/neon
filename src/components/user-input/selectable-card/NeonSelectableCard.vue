<template>
  <neon-card
    :class="[
      `neon-selectable-card--${color}`,
      { 'neon-selectable-card--selected': modelValue },
      { 'neon-selectable-card--disabled': disabled },
    ]"
    :horizontal-borders="false"
    :no-break="true"
    :size="size"
    class="neon-selectable-card"
    orientation="horizontal"
    tabindex="0"
    @click="clicked()"
    @keydown.enter.stop.prevent="clicked()"
    @keydown.space.stop.prevent="clicked()"
  >
    <neon-card-header>
      <neon-switch
        :color="color"
        :disabled="disabled"
        :model-value="modelValue"
        :size="size"
        class="neon-selectable-card__checkbox"
        switch-style="checkbox"
        @update:modelValue="clicked()"
      />
    </neon-card-header>
    <!-- @slot the card sections to add to the card after the checkbox -->
    <slot name="default"></slot>
  </neon-card>
</template>

<script lang="ts" src="./NeonSelectableCard.ts" />
