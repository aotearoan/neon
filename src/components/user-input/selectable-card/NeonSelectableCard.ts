import { defineComponent, ref } from 'vue';
import { NeonSize } from '@/common/enums/NeonSize';
import { NeonFunctionalColor } from '@/common/enums/NeonFunctionalColor';
import NeonCard from '@/components/layout/card/NeonCard.vue';
import NeonCardHeader from '@/components/layout/card/header/NeonCardHeader.vue';
import NeonSwitch from '@/components/user-input/switch/NeonSwitch.vue';
import { NeonResponsive } from '@/common/enums/NeonResponsive';

/**
 * <p>
 * <strong>NeonSelectableCard</strong> is a horizontal card that is selectable by clicking on it. It can be used in
 * lists to easily allow users to select multiple items, providing richer content & a larger footprint for clicking on.
 * </p>
 */
export default defineComponent({
  name: 'NeonSelectableCard',
  components: {
    NeonCard,
    NeonCardHeader,
    NeonSwitch,
  },
  props: {
    /**
     * The selected state of the card
     */
    modelValue: { type: Boolean, required: true },
    /**
     * The size of the card.
     */
    size: { type: String as () => NeonSize, default: NeonSize.Small },
    /**
     * The selection color. This will be used for the checkbox as well as borders & the background color
     */
    color: { type: String as () => NeonFunctionalColor, default: NeonFunctionalColor.Brand },
    /**
     * The disabled state of the card.
     */
    disabled: { type: Boolean, default: false },
  },
  emits: [
    /**
     * Emitted when the user clicks on the card, toggling the selection state.
     * @type {boolean} the current selection state of the card.
     */
    'update:modelValue',
  ],
  setup(props, { emit }) {
    const active = ref<boolean>(false);
    const clicked = () => {
      if (!props.disabled) {
        emit('update:modelValue', !props.modelValue);
      }
    };

    return {
      active,
      clicked,
      NeonResponsive,
    };
  },
});
