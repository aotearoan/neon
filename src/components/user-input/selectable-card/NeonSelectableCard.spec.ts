import type { RenderResult } from '@testing-library/vue';
import { fireEvent, render } from '@testing-library/vue';
import NeonSelectableCard from './NeonSelectableCard.vue';
import { NeonFunctionalColor } from '@/common/enums/NeonFunctionalColor';
import { NeonSize } from '@/common/enums/NeonSize';

describe('NeonSelectableCard', () => {
  let harness: RenderResult;

  beforeEach(() => {
    harness = render(NeonSelectableCard, {
      props: { modelValue: false },
      slots: { default: '<p>test</p>' },
    });
  });

  it('matches snapshot', () => {
    // given
    const { html } = harness;
    // when / then
    expect(html()).toMatchSnapshot();
  });

  it('renders selected', async () => {
    // given
    const { container, rerender } = harness;
    // when
    await rerender({ modelValue: true });
    // then
    expect(container.querySelector('.neon-selectable-card--selected')).toBeDefined();
  });

  it('toggles selected on Enter', async () => {
    // given
    const { container, emitted } = harness;
    // when
    const card = container.querySelector('.neon-selectable-card') as HTMLElement;
    await fireEvent.keyDown(card, { key: 'Enter', code: 'Enter' });
    // then
    expect(emitted()['update:modelValue'][0]).toEqual([true]);
  });

  it('toggles selected on Space', async () => {
    // given
    const { container, emitted } = harness;
    // when
    const card = container.querySelector('.neon-selectable-card') as HTMLElement;
    await fireEvent.keyDown(card, { key: 'Space', code: 'Space' });
    // then
    expect(emitted()['update:modelValue'][0]).toEqual([true]);
  });

  it('toggles selected on when clicked', async () => {
    // given
    const { container, emitted } = harness;
    // when
    const card = container.querySelector('.neon-selectable-card') as HTMLElement;
    await fireEvent.click(card);
    // then
    expect(emitted()['update:modelValue'][0]).toEqual([true]);
  });

  it('does not toggle selected on when disabled', async () => {
    // given
    const { container, emitted, rerender } = harness;
    // when
    await rerender({ disabled: true });
    const card = container.querySelector('.neon-selectable-card--disabled') as HTMLElement;
    await fireEvent.click(card);
    // then
    expect(emitted()['update:modelValue']).toBeUndefined();
  });

  it('renders color', async () => {
    // given
    const { container, rerender } = harness;
    // when / then
    await rerender({ color: NeonFunctionalColor.Info });
    expect(container.querySelector('.neon-selectable-card--info')).toBeDefined();
  });

  it('renders size', async () => {
    // given
    const { container, rerender } = harness;
    // when / then
    await rerender({ size: NeonSize.Medium });
    expect(container.querySelector('.neon-selectable-card--m')).toBeDefined();
  });

  it('toggles selected off when clicked', async () => {
    // given
    const { container, emitted, rerender } = harness;
    // when
    await rerender({ modelValue: true });
    const card = container.querySelector('.neon-selectable-card') as HTMLElement;
    await fireEvent.click(card);
    // then
    expect(emitted()['update:modelValue'][0]).toEqual([false]);
  });

  it('renders default slot contents', () => {
    // given
    const { html } = harness;
    // when / then
    expect(html()).toMatch('<p>test</p>');
  });
});
