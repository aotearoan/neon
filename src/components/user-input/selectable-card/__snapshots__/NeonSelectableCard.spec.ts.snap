// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NeonSelectableCard matches snapshot 1`] = `
<div class="neon-card--horizontal neon-card--s neon-card--horizontal-breakpoint-mobile-l neon-card--horizontal-breakpoint-no-break neon-card neon-selectable-card--brand neon-selectable-card neon-selectable-card--brand neon-selectable-card" tabindex="0">
  <!-- @slot card contents -->
  <div class="neon-card-header">
    <!-- @slot card header content --><label aria-checked="false" aria-disabled="false" class="neon-switch--checkbox neon-switch--s neon-switch--brand neon-switch--right neon-switch no-style neon-selectable-card__checkbox" role="checkbox" tabindex="-1">
      <div class="neon-icon--name-check neon-icon--inverse neon-icon neon-switch__checkbox neon-switch__checkbox"><svg width="100%" height="100%" viewBox="0 0 24 24" style="fill-rule:evenodd;clip-rule:evenodd;stroke-linejoin:round;stroke-miterlimit:2;">
          <path class="neon-svg--fill" d="M22.596,6.529c0.26,-0.278 0.404,-0.645 0.404,-1.025c0,-0.825 -0.679,-1.504 -1.505,-1.504c-0.381,0 -0.748,0.144 -1.026,0.404l-11.981,11.971l-4.957,-4.953c-0.278,-0.259 -0.645,-0.404 -1.026,-0.404c-0.826,0 -1.505,0.679 -1.505,1.504c-0,0.381 0.144,0.747 0.404,1.025l6.02,6.016c0.584,0.583 1.544,0.583 2.128,0l13.044,-13.034Z" style="fill-rule:nonzero;"></path>
        </svg>
      </div>
      <!--v-if--><input class="neon-switch__input neon-selectable-card__checkbox" type="checkbox">
      <!--v-if-->
    </label>
  </div><!-- @slot the card sections to add to the card after the checkbox -->
  <p>test</p>
</div>
`;
